File: auto_query_genius/evaluation/csv_parser.py

"""
CSV Parser Module

This module provides functionality for parsing CSV benchmark datasets.
"""

import csv
import io
import json
from typing import List, Dict, Any, Optional


def parse_csv_dataset(csv_content: str) -> List[Dict[str, Any]]:
    """
    Parse a CSV benchmark dataset.
    
    Args:
        csv_content (str): The CSV content to parse
        
    Returns:
        List[Dict[str, Any]]: A list of job description items with ground truth keywords
        
    Raises:
        ValueError: If the CSV is missing required columns or has invalid data
    """
    # Parse CSV
    reader = csv.DictReader(io.StringIO(csv_content))
    
    if not reader.fieldnames:
        raise ValueError("CSV file has no header row")
    
    # Debug header information
    print(f"CSV Headers found: {reader.fieldnames}")
    
    # Normalize header names to lowercase
    header_map = {name: name.lower().strip() for name in reader.fieldnames}
    
    # Check for required columns in various formats
    id_fields = ['id', 'job_id', 'company_name', 'position_id', 'index']
    description_fields = ['description', 'text', 'job_description', 'content', 'job posting']
    keyword_fields = [
        'keywords', 'ground_truth', 'groundtruth', 'model_response',
        'expected_keywords', 'manual_keywords', 'annotated_keywords',
        'skills', 'tags'  # Add more possible keyword fields
    ]
    
    # Validate that at least one description field is present
    has_description_field = any(field in header_map.values() for field in description_fields)
    
    if not has_description_field:
        print(f"Warning: CSV headers don't match expected format. Available headers: {list(header_map.values())}")
        # Try to guess which column might contain the job description
        text_column_candidates = [col for col in header_map.values() if 'text' in col or 'description' in col or 'content' in col or 'job' in col]
        if text_column_candidates:
            print(f"Using '{text_column_candidates[0]}' as the description field")
            description_fields.append(text_column_candidates[0])
        else:
            raise ValueError(f"CSV file must contain one of these description columns: {', '.join(description_fields)}")
    
    # Parse each row
    items = []
    for i, row in enumerate(reader):
        try:
            # Map row keys to lowercase for consistent access
            row_lower = {k.lower().strip(): v for k, v in row.items()}
            
            # Extract job ID from available fields or use row index
            job_id = None
            for field in id_fields:
                if field in row_lower and row_lower[field].strip():
                    job_id = row_lower[field].strip()
                    break
            
            if not job_id:
                job_id = f"job-{i+1}"
            
            # Extract description from available fields
            description = ""
            for field in description_fields:
                if field in row_lower and row_lower[field].strip():
                    description += row_lower[field].strip() + " "
            
            # Special handling for the Hugging Face dataset format
            if 'job posting' in row_lower and row_lower['job posting'].strip():
                description = row_lower['job posting'].strip()
            
            description = description.strip()
            if not description:
                continue  # Skip rows with empty descriptions
            
            # Add position title to description if available
            if 'position_title' in row_lower and row_lower['position_title'].strip():
                description = f"{row_lower['position_title'].strip()}: {description}"
            elif 'title' in row_lower and row_lower['title'].strip():
                description = f"{row_lower['title'].strip()}: {description}"
            elif 'job title' in row_lower and row_lower['job title'].strip():
                description = f"{row_lower['job title'].strip()}: {description}"
            
            # Generate synthetic keywords from the description if no keyword fields found
            # This is helpful for datasets like the Hugging Face one that doesn't have explicit keywords
            keywords = []
            
            # First try to parse from existing keyword fields
            keywords_found = False
            for field in keyword_fields:
                if field in row_lower and row_lower[field].strip():
                    try:
                        # Parse as JSON if possible
                        if row_lower[field].startswith('{') or row_lower[field].startswith('['):
                            keyword_data = json.loads(row_lower[field])
                            # Process JSON format
                            if isinstance(keyword_data, list):
                                for kw in keyword_data:
                                    if isinstance(kw, dict):
                                        keywords.append({
                                            'keyword': kw.get('keyword', '') or kw.get('term', ''),
                                            'frequency': kw.get('frequency', 1) or kw.get('count', 1)
                                        })
                                    elif isinstance(kw, str):
                                        keywords.append({
                                            'keyword': kw,
                                            'frequency': 1
                                        })
                            elif isinstance(keyword_data, dict):
                                for key, value in keyword_data.items():
                                    if key and key != 'N/A':
                                        frequency = 1
                                        if isinstance(value, (int, float)):
                                            frequency = value
                                        keywords.append({
                                            'keyword': key,
                                            'frequency': frequency
                                        })
                        else:
                            # Parse as comma-separated list
                            keyword_list = row_lower[field].split(',')
                            for kw_entry in keyword_list:
                                kw_entry = kw_entry.strip()
                                if kw_entry:
                                    if ':' in kw_entry:
                                        kw, freq = kw_entry.split(':', 1)
                                        try:
                                            freq = int(freq.strip())
                                        except ValueError:
                                            freq = 1
                                        keywords.append({
                                            'keyword': kw.strip(),
                                            'frequency': freq
                                        })
                                    else:
                                        keywords.append({
                                            'keyword': kw_entry,
                                            'frequency': 1
                                        })
                        keywords_found = True
                        break
                    except Exception as e:
                        print(f"Error parsing keywords for job {job_id} from field {field}: {e}")
                        continue
            
            # If no keywords were found, extract important terms from the description
            if not keywords_found:
                import re
                from collections import Counter
                
                # Extract potential keywords from the description
                # Remove common words and keep only potentially meaningful terms
                words = re.findall(r'\b[A-Za-z][A-Za-z0-9+#\.]{2,}\b', description.lower())
                stopwords = {'and', 'the', 'to', 'of', 'a', 'in', 'for', 'with', 'on', 'at', 'from', 'by', 'about', 
                             'as', 'into', 'like', 'through', 'after', 'over', 'between', 'out', 'against', 'during', 
                             'without', 'before', 'under', 'around', 'among', 'our', 'we', 'us', 'you', 'they', 'them', 
                             'their', 'this', 'that', 'these', 'those', 'will', 'have', 'has', 'had', 'not', 'are', 
                             'was', 'were', 'been', 'being', 'have', 'has', 'had', 'do', 'does', 'did', 'can', 'could', 
                             'should', 'would', 'may', 'might', 'must', 'shall', 'should'}
                
                filtered_words = [word for word in words if word not in stopwords and len(word) > 3]
                word_counts = Counter(filtered_words)
                
                # Get top 15 most common words as synthetic keywords
                for word, count in word_counts.most_common(15):
                    keywords.append({
                        'keyword': word,
                        'frequency': count
                    })
                
                # Also try to extract multi-word technical terms using regex patterns
                tech_terms = re.findall(r'\b[A-Za-z][A-Za-z0-9]*[\s\-\.][A-Za-z0-9]+\b', description)
                for term in tech_terms[:10]:  # Get up to 10 multi-word terms
                    clean_term = term.lower().strip()
                    if clean_term and len(clean_term) > 5 and all(w not in stopwords for w in clean_term.split()):
                        keywords.append({
                            'keyword': clean_term,
                            'frequency': 1
                        })
            
            # Only include job if it has description and keywords
            if description and keywords:
                items.append({
                    'id': job_id,
                    'text': description,
                    'keywords': keywords
                })
                
        except Exception as e:
            print(f"Error processing row {i}: {e}")
            continue
    
    print(f"Successfully parsed {len(items)} items from CSV")
    if items:
        print(f"Sample item: {items[0]['id']} with {len(items[0]['keywords'])} keywords")
        print(f"Sample keywords: {items[0]['keywords'][:3]}")
    
    return items

File: src/components/evaluation/utils/csvParser.ts

import Papa from "papaparse";
import { EvaluationDataItem, KeywordItem } from "../types";

// Helper function to extract keywords from JSON string in model_response
export const extractKeywordsFromModelResponse = (modelResponse: string): KeywordItem[] => {
  try {
    console.log("Trying to parse model_response:", modelResponse);
    
    // Check if it's already a valid JSON string format
    if (modelResponse && typeof modelResponse === 'string') {
      // Clean up the string - some CSV exports might have formatting issues
      let cleanedJson = modelResponse.trim();
      
      // Parse the JSON
      const parsedData = JSON.parse(cleanedJson);
      
      // Look for keywords in common fields
      if (parsedData) {
        // If it has Core Responsibilities or similar fields, extract keywords from there
        const possibleKeywordSources = [
          parsedData.keywords,
          parsedData.Core_Responsibilities,
          parsedData["Core Responsibilities"],
          parsedData.skills,
          parsedData.tags
        ];
        
        for (const source of possibleKeywordSources) {
          if (Array.isArray(source)) {
            return source.map(kw => typeof kw === 'string' ? { keyword: kw, frequency: 1 } : kw);
          } else if (source && typeof source === 'string') {
            return source.split(',').map(kw => ({ keyword: kw.trim(), frequency: 1 }));
          }
        }
        
        // If we found structured data but no keywords array, look for key-value pairs
        if (typeof parsedData === 'object') {
          // Extract keys from the object as keywords
          const extractedKeywords = [];
          for (const [key, value] of Object.entries(parsedData)) {
            if (key !== 'N/A' && key !== 'n/a' && key.length > 2) {
              if (typeof value === 'string' && value !== 'N/A') {
                extractedKeywords.push({
                  keyword: key.trim(),
                  frequency: 1,
                  category: value
                });
              } else {
                extractedKeywords.push({
                  keyword: key.trim(),
                  frequency: 1
                });
              }
            }
          }
          if (extractedKeywords.length > 0) {
            return extractedKeywords;
          }
        }
      }
    }
    
    // If we couldn't extract structured data, just extract the content as plain text
    return [{ keyword: "No valid keywords found", frequency: 0 }];
  } catch (error) {
    console.error("Error parsing model_response:", error, "Original:", modelResponse);
    
    // Fall back to basic string parsing if JSON parsing fails
    if (modelResponse && typeof modelResponse === 'string') {
      // Try to extract anything that looks like a keyword
      const keywordMatches = modelResponse.match(/["']([^"']+)["']/g);
      if (keywordMatches && keywordMatches.length > 0) {
        return keywordMatches.map(match => ({
          keyword: match.replace(/["']/g, '').trim(),
          frequency: 1
        }));
      }
    }
    
    return [{ keyword: "Failed to parse", frequency: 0 }];
  }
};

export const parseCSV = (text: string): EvaluationDataItem[] => {
  try {
    // Parse CSV with header
    const results = Papa.parse(text, { 
      header: true, 
      skipEmptyLines: true,
      transformHeader: (header) => header.trim().toLowerCase()
    });
    
    if (!results.data || !Array.isArray(results.data) || results.data.length === 0) {
      throw new Error("Invalid CSV format or empty file");
    }

    console.log("CSV parsing result:", {
      rowCount: results.data.length,
      firstRow: results.data[0],
      fields: Object.keys(results.data[0] || {})
    });

    // Transform the CSV data into our expected format
    return results.data
      .filter((row: any) => row && typeof row === 'object')
      .map((row: any, index) => {
        // Debug each row
        console.log(`Processing row ${index}:`, row);
        
        // Extract ID from various possible fields
        let id = row.id || row.company_name || row.position_id || `item-${index}`;
        
        // Extract description from various possible fields
        let description = "";
        if (row.job_description || row.description || row.text || row.content) {
          description = row.job_description || row.description || row.text || row.content || "";
          
          // If there's a position_title, prepend it to the description
          if (row.position_title || row.title) {
            description = `${row.position_title || row.title}: ${description}`;
          }
        } else {
          console.warn(`Row ${index} missing description:`, row);
          return null;
        }
        
        // Extract ground truth keywords from model_response
        let groundTruth: KeywordItem[] = [];
        
        // Check for model_response field first
        if (row.model_response && typeof row.model_response === 'string') {
          groundTruth = extractKeywordsFromModelResponse(row.model_response);
          console.log(`Extracted ${groundTruth.length} keywords from model_response for row ${index}`);
        }
        
        // If no keywords found in model_response, try other fields
        if (groundTruth.length === 0) {
          // Try different potential field names for keywords
          const keywordFields = [
            'keywords', 'ground_truth', 'groundtruth', 
            'ground_truth_keywords', 'actual_keywords', 'expected_keywords',
            'manual_keywords', 'annotated_keywords'
          ];
          
          // Check each possible field name
          for (const field of keywordFields) {
            if (row[field] && typeof row[field] === 'string') {
              try {
                // Try to parse as JSON
                if (row[field].trim().startsWith('[') || row[field].trim().startsWith('{')) {
                  const parsed = JSON.parse(row[field]);
                  
                  if (Array.isArray(parsed)) {
                    // Handle array of strings or objects
                    groundTruth = parsed.map(item => {
                      if (typeof item === 'string') {
                        return { keyword: item.trim(), frequency: 1 };
                      } else if (item && typeof item === 'object') {
                        return {
                          keyword: (item.keyword || item.term || item.text || "").trim(),
                          frequency: item.frequency || item.count || item.weight || 1
                        };
                      }
                      return null;
                    }).filter(Boolean);
                    break;
                  } else if (parsed && typeof parsed === 'object') {
                    // Handle object with key-value pairs
                    groundTruth = Object.entries(parsed).map(([key, value]) => ({
                      keyword: key.trim(),
                      frequency: typeof value === 'number' ? value : 1
                    }));
                    break;
                  }
                } else {
                  // Try parsing as comma-separated list
                  groundTruth = row[field].split(',').map((keyword: string) => {
                    const [term, countStr] = keyword.split(':');
                    return {
                      keyword: term.trim(),
                      frequency: parseInt(countStr?.trim() || "1", 10) || 1
                    };
                  }).filter((item: any) => item.keyword);
                  break;
                }
              } catch (error) {
                console.warn(`Failed to parse ${field} for row ${index}:`, error);
                // Continue to the next field if parsing fails
              }
            }
          }
        }
        
        // If still no keywords found, extract some from the description for testing
        if (groundTruth.length === 0) {
          console.warn(`No ground truth keywords found for row ${index}. Creating fallback keywords.`);
          
          // For testing, create a minimal set of dummy keywords from the description
          const words = description.split(/\s+/).filter(word => 
            word.length > 4 && !/^\d+$/.test(word) && !['the', 'this', 'that', 'with', 'from', 'your'].includes(word.toLowerCase())
          ).slice(0, 5);
          
          groundTruth = words.map(word => ({
            keyword: word.toLowerCase().replace(/[^\w]/g, ''),
            frequency: 1
          }));
          
          console.log(`Created fallback keywords for testing:`, groundTruth);
        }

        return {
          id,
          description,
          groundTruth: groundTruth.filter(kw => kw && kw.keyword && kw.keyword.trim() !== "")
        };
      })
      .filter(Boolean); // Remove null items
  } catch (error) {
    console.error("Error parsing CSV:", error);
    throw error;
  }
};

// Parse JSON data files
export const parseJSON = (content: string): EvaluationDataItem[] => {
  try {
    const jsonData = JSON.parse(content);
    
    // Validate JSON structure
    if (!Array.isArray(jsonData) || jsonData.length === 0) {
      throw new Error("Invalid dataset format. Expecting an array of evaluation items.");
    }

    // Validate and transform items if needed
    return jsonData
      .filter((item: any) => item && typeof item === 'object')
      .map((item: any, index) => {
        if (!item.description) {
          console.warn("Missing description in item at index", index);
          return null;
        }
        
        let groundTruth = [];
        if (Array.isArray(item.groundTruth)) {
          groundTruth = item.groundTruth
            .filter((kw: any) => kw)
            .map((kw: any) => {
              if (typeof kw === 'string') {
                return { keyword: kw, frequency: 1 };
              } else if (kw && typeof kw === 'object') {
                return {
                  keyword: kw.keyword || "",
                  frequency: typeof kw.frequency === 'number' ? kw.frequency : 1
                };
              }
              return null;
            })
            .filter(Boolean);
        } else {
          console.warn("Missing or invalid groundTruth in item at index", index);
          return null;
        }
        
        return {
          id: item.id || `item-${index}`,
          description: item.description,
          groundTruth
        };
      })
      .filter(Boolean);
  } catch (error) {
    console.error("Error parsing JSON:", error);
    throw error;
  }
};
